docs/reference/wasm.md

```diff
@@ -0,0 +1,241 @@
+---
+title: 'wasm'
+weight: 110
+description: 'Wasm API exposed to guests'
+---
+
+Inspektor Gadget exposes some functions to wasm modules implemented in gadgets.
+We provide a Golang wrapper for this functionality, but these functions can be
+used directly from any programming language that can be compiled to wasm.
+
+## Data types
+
+Data types passed to the API are encoded using 64 bits. Scalar types like
+integers, booleans and floats are casted directly to 64 bits integer and passed
+using the stack.
+
+Strings and byte arrays are stored in the wasm module's memory. A 64 bits
+integer is used to represent a pointer to them, the higher 32 bits contains the
+length and the lower 32 the memory address.
+
+## API
+
+### Log
+
+#### `gadgetLog(u32 level, string msg)`
+
+Print a log message using the gadget's logger instance.
+
+Parameters:
+
+- `level` (u32): Log level:
+  - 0: Error
+  - 1: Warn
+  - 2: Info
+  - 3: Debug
+  - 4: Trace
+- `msg` (string): Message to print
+
+Return value:
+- None
+
+### Datasources
+
+#### `newDataSource(string name)`
+
+Creates a new data source.
+
+Parameters:
+- `name` (string): Data source name
+
+Return value:
+- (u32): Handle to the created data source on success, 0 on error.
+
+#### `getDataSource(string name) u32`
+
+Get a handle to a data source.
+
+Parameters:
+- `name` (string): Datasource's name
+
+Return value:
+- (u32) Handle to the data source on success, 0 on error.
+
+#### `dataSourceSubscribe(u32 ds, u32 type, u32 prio, u64 cb)`
+
+Subscribe to events emitted by a data source.
+
+This mechanism requires the wasm module to export a `dsCallback` that is called
+by the host when an event is emitted:
+
+`dsCallback(u64 cbID, u32 ds, u32 data)`
+- `cbID`: Callback ID
+- `ds`: Datasource handle
+- `data`: Data handle
+
+Parameters:
+- `ds` (u32): Datasource handle (as returned by `getDataSource` or `newDataSource`)
+- `type` (u32): Subscription type: (0: Data, 1: Array, 2: Packet)
+- `priority` (u32): Priority of the subscription. Lower means higher priority.
+- `cb` (u64): Callback ID passed to `dsCallback`
+
+Return value:
+- 0 on sucess, 1 in case of error.
+
+#### `dataSourceGetField(u32 ds, string name) u32`
```

Comment:
Maybe we should return a u64 and use the upper bits to carry the kind to make it implicit for `fieldAccessor*`.